applications/ShapeOptimizationApplication/python_scripts/algorithms/algorithm_gradient_projection.py

```diff
@@ -148,73 +148,95 @@ def __analyzeShape(self):
 
     # --------------------------------------------------------------------------
     def __computeShapeUpdate(self):
-        for name, mapper in self.mappers.items():
+        for mapper in self.mappers.values():
             mapper.Update()
             mapper.InverseMap(KSO.DF1DX, KSO.DF1DX_MAPPED)
 
-            for constraint in self.constraints:
-                con_id = constraint["identifier"].GetString()
-                gradient_variable = self.constraint_gradient_variables[con_id]["gradient"]
-                mapped_gradient_variable = self.constraint_gradient_variables[con_id]["mapped_gradient"]
-                mapper.InverseMap(gradient_variable, mapped_gradient_variable)
+        for constraint in self.constraints:
+            con_id = constraint["identifier"].GetString()
+            gradient_variable = self.constraint_gradient_variables[con_id]["gradient"]
+            mapped_gradient_variable = self.constraint_gradient_variables[con_id]["mapped_gradient"]
+            mapper.InverseMap(gradient_variable, mapped_gradient_variable)
 
-            self.__computeControlPointUpdate()
+        self.__computeControlPointUpdate()
 
+        for mapper in self.mappers.values():
             mapper.Map(KSO.CONTROL_POINT_UPDATE, KSO.SHAPE_UPDATE)
-            self.model_part_controller.DampNodalVariableIfSpecified(KSO.SHAPE_UPDATE)
+
+        self.model_part_controller.DampNodalVariableIfSpecified(KSO.SHAPE_UPDATE)
 
     # --------------------------------------------------------------------------
     def __computeControlPointUpdate(self):
         """adapted from https://msulaiman.org/onewebmedia/GradProj_2.pdf"""
         g_a, g_a_variables = self.__getActiveConstraints()
-
-        for name, design_surface in self.design_surfaces.items():
-            KM.Logger.PrintInfo("ShapeOpt", "Assemble vector of objective gradient.")
-            nabla_f = KM.Vector()
-            s = KM.Vector()
-            optimization_utilities.AssembleVector(design_surface, nabla_f, KSO.DF1DX_MAPPED)
-
-            if len(g_a) == 0:
-                KM.Logger.PrintInfo("ShapeOpt", "No constraints active, use negative objective gradient as search direction.")
+        if len(g_a) == 0:
+            KM.Logger.PrintInfo("ShapeOpt", "No constraints active, use negative objective gradient as search direction.")
+            for design_surface in self.design_surfaces.values():
+                s = KM.Vector()
```

Comment:
`nabla_f` does not yet exist in this if statement but is already used to compute s in case of no active constraints.

Reply:
True ! changed !